### PR TITLE
Propose draft text for https://github.com/mozilla/pkipolicy/issues/170

### PR DESCRIPTION
**Pull request by sleevi**

### CODE REVIEW

File: rootstore/policy.md [hunk 1]
```diff
@@ -364,12 +364,137 @@ set:
...
+When RSA keys are encoded in a SubjectPublicKeyInfo structure, the algorithm
+field MUST consist of an rsaEncryption OID (1.2.840.113549.1.1.1) with a NULL
+parameter, as specified by [RFC 8017, Appendix A.1](https://tools.ietf.org/html/rfc8017#appendix-A.1)
+and [RFC 3279, Section 2.3.1](https://tools.ietf.org/html/rfc3279#section-2.3.1).
+The encoded AlgorithmIdentifier for an RSA key MUST match the
+following hex-encoded bytes:
+`300d06092a864886f70d0101010500`.
+
+CAs MUST NOT use the id-RSASSA-PSS OID (1.2.840.113549.1.1.10) within a
+SubjectPublicKeyInfo to represent a RSA key.
```
Comment (briansmith) [inline rootstore/policy.md]: Why not? I think it would be useful to have PSS-only public keys in certificates.

**Comment by briansmith** [comment]
I checked all of these except the RSA SHA-1 encoding against [webpki](https://github.com/briansmith/webpki) and they match what webpki uses.

**Comment by sleevi** [comment]
Useful != Usable.

In the case of RSA-PSS SPKIs, they are, as you know, incompatible with
every major implementation, including Firefox. The semantics of the
parameters - and the enforcement of them - is equally problematic, as
captured in Section 3.1 of RFC 4055.

If we were to imagine there was a need and/or desire for them, then as
noted, using a different OID would be far more beneficial.

But, as a practical matter, they don’t work and aren’t enforced.

**Comment by briansmith** [comment]
> In the case of RSA-PSS SPKIs, they are, as you know, incompatible with every major implementation, including Firefox.

Imagine we wanted to impose a requirement on all CAs; would we avoid imposing the new requirement solely because no CA had yet complied with the requirement? I don't think so. We should treat browsers the same way. So, I don't think this is a valid argument as stated.

> The semantics of the parameters - and the enforcement of them - is equally problematic, as
captured in Section 3.1 of RFC 4055.

I agree that it seems like RFC 4055 and RFC 5756 don't fully capture all the requirements that should be imposed on the use of PSS public keys and so I think it makes sense to defer the specification of how PSS public keys may be used until an actual good specification for PSS keys is used. (However, I don't see anything problematic about RFC 4055/5756 requirements *for PSS* that would be incompatible with the stronger requirements that should be imposed.)

is the PSS subjectPublicKeyInfo the only case where this change is imposing ("MUST") requirements for a particular encoding in situations where the RFCs only recommend ("SHOULD") or allow ("MAY") a particular encoding?

**Comment by sleevi** [comment]
Considering that this is a deliberate choice, similar to P-521 and DSA, and one borne out of complexity in implementation, use, and correctness, and which has demonstrably not been used, I fail to understand your objection. Especially given the past support for prohibiting those things for similar reasons as this. It’s important to note that this doesn’t prohibit the use of RSA-PSS, it only prevents a more restrictive, but ill-defined and unsupported, encoding.

As for other MUST levels - clearly, the requirement to have hash/key agreement in the case of ECC is a MUST requirement being imposed on a SHOULD. One could also argue the same of key sizes. This also explicitly restricts to the NIST curves, although that too is a policy matter that was previously debated and agreed upon.

**Pull request merged**